facenet_sandberg/generate_pairs.py

```diff
@@ -4,139 +4,124 @@
 
 import os
 import random
-from typing import List, Tuple
-
+import sys
+import argparse
 import numpy as np
+from typing import List, Tuple, cast
+
 
-names = [
-    d for d in os.listdir(image_dir) if os.path.isdir(
-        os.path.join(image_dir, d))]
+Mismatch = Tuple[str, int, str, int]
+Match = Tuple[str, int, int]
 
 
-def split_people_into_sets(image_dir: str, k_num_sets: int) -> List[List[str]]:
-    names = [
-        d for d in os.listdir(image_dir) if os.path.isdir(
-            os.path.join(
-                image_dir,
-                d))]
+def write_pairs(fname: str,
+                match_folds: List[List[Match]],
+                mismatch_folds: List[List[Mismatch]],
+                k_num_sets: int,
+                total_matches_mismatches: int) -> None:
+    file_contents = f'{k_num_sets}\t{total_matches_mismatches}\n'
+    for match_fold, mismatch_fold in zip(match_folds, mismatch_folds):
+        for match in match_fold:
+            file_contents += f'{match[0]}\t{match[1]}\t{match[2]}\n'
```

Comment:
Given that we're building up the file contents here and then writing them to the file, why not write the file incrementally to avoid a potentially unbounded memory overhead? Also makes the code somewhat neater since you don't need the string concatenation.

Reply:
I wrote it this way to avoid constantly writing to disk. I will change it in the next push

Reply:
I did a bit of testing and actually it turns out that you're right and this approach is faster (see below). Building the string in memory gives a speedup of roughly 2x.

However, you can let the OS/stdlib take care of optimizing the disk access by passing a buffer size argument to `open` (e.g. `open('file.txt', 'w', 512)`) in which case the difference starts disappearing.

```
import random

def concat_write():
    s = ""
    for x in range(100000):
        s += "%s\n" % x

    with open('foo%f' % random.random(), 'w') as fobj:
        fobj.write(s)


def incremental_write():
    with open('bar%f' % random.random(), 'w') as fobj:
        for x in range(100000):
            s = "%s\n" % x
            fobj.write(s)


from timeit import timeit
print(timeit('incremental_write()', globals=globals(), number=100))
print(timeit('concat_write()', globals=globals(), number=100))
```

Reply:
Ok cool, I will let the OS take care of buffering, using the default io buffer size. It is in the next commit.